### PR TITLE
SMTH-52: Quest Status Tracking

### PR DESCRIPTION
**Pull request by dbarton-aoe**
**Description**
Refactor of a number of systems and the introduction of the quest management system including the quest database, quest status tracker and the quest log player component.

[Quest Management Video](https://github.com/aoe-games/SMTH/files/5543058/2020-11-15_QuestStateManagement.zip)

**Details**
* QuestManager has a global event used to pass along user quest selection
* GameServices class is a singleton used to load and give access to the various services in the game
* QuestEntryCtrl and QuestEntryView created to manage the map's entry points based on quest state changes and sends selection messages through the QuestManager for the QuestCtrl to pickup and action encounters setup
* Created World and QuestInterface scenes
* World scene has quest entry points and the QuestInterface now handles the showing and hiding of the quest interfaces
* QuestInterfaceScene now loads the World scene once the Splash screen has been unloaded
* Splash screen now creates and initializes the GameServices object and the player
* LocalPlayer now has a PlayerQuestLog player component housing a QuestStatusDatabase
* QuestStatusDatabase houses a collection of QuestStatusData objects used to track the status of the player's quests
* Created and added all quest data for the quests on the world map to the quest database 
* Reformatted Classes:
  * QuestStatusData changed to State
  * Reformatted the QuestCtrl class
  * Reformatted the QuestStateData and QuestStateDatabase classes

World_2FinishedQuests

### CODE REVIEW

File: UnityProject/Assets/Scripts/Quest/QuestCtrl.cs [hunk 1]
```diff
@@ -1,135 +1,164 @@
...
+  protected EncounterObserver m_encounterObserver = null;
+  [SerializeField]
+  protected RectTransform m_encounterObserverView = null;
+  [SerializeField]
+  protected ActorFactory m_actorFactory = null;
+
+  [SerializeField]
+  protected List<PartyData> m_debug_parties = new List<PartyData>();
+
+  protected List<PartyData> m_parties = new List<PartyData>();
```
Comment (richardlalancette) [inline UnityProject/Assets/Scripts/Quest/QuestCtrl.cs:28]: Class getting a little big.
Maybe see if you can split the functionality into smaller classes?

File: UnityProject/Assets/Scripts/Quest/QuestData/QuestStateData.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+/// </summary>
+public class QuestStateData
+{
+  public enum Status : byte { Unavailable, Available, InProgress, Complete  }
+
+  public string ID { get; protected set; }
+  public Status QuestStatus { get; protected set; }
+  public DateTime CompletionTime { get; protected set; }
+
+  public bool HasCompletionTimeExpired
```
Comment (richardlalancette) [inline UnityProject/Assets/Scripts/Quest/QuestData/QuestStateData.cs:18]: Nice ;)

File: UnityProject/Assets/Scripts/Scenes/Splash.cs [hunk 3]
```diff
@@ -30,19 +33,34 @@ public override IEnumerator UnloadScene(ProgressDelegate progressDelegate = null
...
 
     Player newPlayer = Instantiate(m_playerPrototype);
     newPlayer.Load(onPlayerLoaded);
+
+    // load services
+    bool areServicesLoaded = false;
+    GameServices.Services = m_services;
+    GameServices.Services.LoadCompletedCallback += (exceptions) => 
+      {
+        areServicesLoaded = true;
```
Comment (richardlalancette) [inline UnityProject/Assets/Scripts/Scenes/Splash.cs:56]: ServicesLoaded

File: UnityProject/Assets/Scripts/Scenes/Splash.cs [hunk 4]
```diff
@@ -30,19 +33,34 @@ public override IEnumerator UnloadScene(ProgressDelegate progressDelegate = null
...
+      };
+
+    GameServices.Services.LoadServices(this);
+
+    yield return new WaitUntil(() => isPlayerLoaded && areServicesLoaded);
```
Comment (richardlalancette) [inline UnityProject/Assets/Scripts/Scenes/Splash.cs:61]: Unusual to see verb in a boolean. No big deal.
Typically verb used in method names.

Comment (richardlalancette) [reply UnityProject/Assets/Scripts/Scenes/Splash.cs:61]: Very small double take :) ahah

**Review by richardlalancette — APPROVED** [review]
Nice work my friend.
I really like how small all your files are. This made it must easier to review and is a good sign they do only one thing :)

**Comment by dbarton-aoe** [comment]
Thank you, @richardlalancette ! Really appreciate you taking the time to have a look and talk things through. Sometimes get lost in my thoughts and having the perspective really helps me see the other angles.

I'll address your feedback and get this pushed in. Thanks again!

File: UnityProject/Assets/Scripts/Quest/QuestCtrl.cs (hunk 1 above)
Comment (dbarton-aoe) [reply UnityProject/Assets/Scripts/Quest/QuestCtrl.cs:28]: Yeah, I see a refactor on the horizon. Will try to revisit in the upcoming work.

**Pull request merged**